### PR TITLE
Add performance tests via rails-perftest (stuck!)

### LINKED ISSUE #1801: Add performance tests for leading pages, like / and /dashboard
https://github.com/publiclab/plots2/blob/8e6ec5a29a7ebb4c873cf5b953468addc1df6547/test/performance/browsing_test.rb#L10

This is documented at https://rubygems.org/gems/rails-perftest (as of Rails 4, which we're moving to shortly)

@icarito - just a note!

### PR DESCRIPTION
**Pull request by bellvat**
* [x] all tests pass -- `rake test:all` **Please see "Issue" below**
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

I have made the following changes -- fixes #1801 :
1. Added perftest gem
2. Added ruby prof gem
3. Uncommented code in browsing_test.rb, and changed syntax of line 9 to: 
 `  test "browsers" do `
4. I have not make changes to the Rakefile yet. 

**Issue:** When I ran `$ bundle exec rake test:benchmark`, which should give us the performance statistics, I kept getting this error:
screen shot 2017-12-08 at 3 39 52 pm

1. I tried to debug this by changing my ruby version, as I read in some posts that it might be that Rails 4.1.6 may not be compatible with Ruby 2.4. I am not as familiar with how rbenv works, and although I was able to install a different version of ruby, I was not able to get rails to work with that version.
2. The same error message came up when I tried running `rails s' to try and start the server

Any guidance would be appreciated! 

Thanks for your help!

### CODE REVIEW

**Comment by jywarren** [comment]
OK, I heard from another contributor (here: https://github.com/publiclab/plots2/issues/1829#issuecomment-350531276) that switching to Ruby 2.3.3 works. We should file an issue to make the app compatible with Ruby 2.4, but for now I think that should resolve your local issue? Are you using RVM or Rbenv, and can you try installing Ruby 2.3.3 for now?

**Comment by jywarren** [comment]
Well cool ! I see your tests do run -- but we see an error:

```
BrowsingTest#test_browsers (29 ms warmup)
        process_time: 120 ms
/usr/local/bundle/gems/ruby-prof-0.16.2/lib/ruby-prof/printers/call_tree_printer.rb:46:in `print': wrong number of arguments (given 2, expected 0..1) (ArgumentError)
```

I think we need `puts` instead of `print`? Let me check the code.

Great work on this in any case!

File: Gemfile [hunk 1]
```diff
@@ -106,3 +106,7 @@ gem 'jbuilder'
 
 # Pin mustermann to Ruby 2.1 compatible
 gem 'mustermann' , '~> 0.4'
+
+#Performance testing
+gem 'rails-perftest'
```
Comment (jywarren) [inline Gemfile]: Could we move this inside the "test" section of the gemfile? Thank you!

**Comment by jywarren** [comment]
Aha! Same issue here: https://github.com/ruby-prof/ruby-prof/issues/203

**Comment by jywarren** [comment]
Hmm, potentially we need to create a directory for the results? 

> It crashes because call_tree_printer needs a directory. Better error message will given when using latest master. (https://github.com/ruby-prof/ruby-prof/issues/215#issuecomment-306163269)

File: Gemfile (hunk 1 above)
Comment (bellvat) [reply Gemfile]: Yes, will move it!

**Comment by jywarren** [comment]
@ryzokuken -- any ideas why we're getting this error? We'd love some help if you have any insights. Thanks!

File: Gemfile (hunk 1 above)
Comment (bellvat) [reply Gemfile]: To clarify, by 'test' section, are you referring to inserting it below 'group :test, :development do', but outside of that that block in the gem file?

Comment (jywarren) [reply Gemfile]: It can go inside that `group :test, :development do` block, thanks!

**Comment by jywarren** [comment]
I left a comment at the `rails-prof` project, but it's odd that I don't see this when running it locally... anyhow, hope we can get some input from them.

**Comment by ryzokuken** [comment]
@jywarren sure, I will take a look.

**Comment by ryzokuken** [comment]
> I tried to debug this by changing my ruby version, as I read in some posts that it might be that Rails 4.1.6 may not be compatible with Ruby 2.4. I am not as familiar with how rbenv works, and although I was able to install a different version of ruby, I was not able to get rails to work with that version.

I believe this error appears because @bellvat seems to be using ruby 2.4. This is a known issue with Rails 4.1 and Ruby 2.4 (https://stackoverflow.com/questions/41504106/ruby-2-4-and-rails-4-stack-level-too-deep-systemstackerror). Infact, Rails 4.1 does not even support Ruby 2.4.

@bellvat please try using Ruby 2.3 (which is what we all use). That'll solve this issue for now. As a long term solution, we'd be migrating to Rails 4.2 soon (Rails 4.2.8 fully supports Ruby 2.4).

File: test/performance/browsing_test.rb [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+require 'test_helper'
+require 'rails/performance_test_help'
 
-#class BrowsingTest < ActionDispatch::PerformanceTest
-  # Refer to the documentation for all available options
-  # self.profile_options = { :runs => 5, :metrics => [:wall_time, :memory]
-  #                          :output => 'tmp/performance', :formats => [:flat] }
+class BrowsingTest < ActionDispatch::PerformanceTest
+   #Refer to the documentation for all available options
+   #self.profile_options = { :runs => 5, :metrics => [:wall_time, :memory]
```
Comment (ryzokuken) [inline test/performance/browsing_test.rb]: Should these still be commented out?

**Comment by ryzokuken** [comment]
@jywarren As of the error that says

```
BrowsingTest#test_browsers (29 ms warmup)
        process_time: 120 ms
/usr/local/bundle/gems/ruby-prof-0.16.2/lib/ruby-prof/printers/call_tree_printer.rb:46:in `print': wrong number of arguments (given 2, expected 0..1) (ArgumentError)
```

it has been fixed in https://github.com/ruby-prof/ruby-prof/commit/837edf447914663d0100e5599ab8c44f81f0eae0. The issue is, the latest version of this gem is 0.16.2, which was released on August 25, 2016. On the other hand, This particular commit was merged on October 19th, 2016. To get over the issue, we should use the gem from GitHub rather than fetching it from RubyGems. http://bundler.io/v1.16/guides/git.html seems to be a good resource regarding this. Basically, we need to change the line to:

```ruby
gem 'ruby-prof', :github => 'ruby-prof'
```

And hopefully that will solve all our problems.

**Comment by bellvat** [comment]
Hi @ryzokuken 

Thank you for taking a look at the issues and advising on possible solutions.

I updated this PR for ruby-prof line in the gemfile to include `:github => 'ruby-prof'` according to the comment above.
Unfortunately, it looks like the error remains.

Not sure what could be the cause..

File: test/performance/browsing_test.rb (hunk 2 above)
Comment (bellvat) [reply test/performance/browsing_test.rb]: I believe tests run with a default setups. If we would like to configure the test runs, we can uncomment the section and specify our own values.

**Comment by ryzokuken** [comment]
A few things: Firstly, the gem still has to be grouped under the `test` group as it had been previously. I didn't mention that in my comment and I'm sorry for the trouble.

Also, after making the changes to the Gemfile, you'd have to run `bundle update` to actually bring those changes in effect. Again, I'm not sure why I missed that as well.

Please run `bundle update` and then try it and let me know if it works. Thanks for your patience.

**Comment by jywarren** [comment]
Thanks both of you! After bundle update, you'll need to re add Gemfile.lock
-- fingers crossed that it works!

File: Gemfile.lock [hunk 3]
```diff
@@ -315,11 +315,11 @@ GEM
     thread_safe (0.3.6)
     tilt (1.4.1)
     timecop (0.9.1)
-    tins (1.15.1)
+    tins (1.16.3)
     tool (0.2.3)
     tzinfo (1.2.4)
       thread_safe (~> 0.1)
-    uglifier (3.2.0)
+    uglifier (4.0.2)
```
Comment (ryzokuken) [inline Gemfile.lock]: @jywarren any idea why it was bumped up by a major version? Wouldn't `4.0.2` be inconsistent with `3.2.0`?

**Comment by jywarren** [comment]
Was version specified in the Gemfile?

**Comment by ryzokuken** [comment]
@bellvat could you try adding `:branch => 'master'` at the end as well? I hope that'll help.

**Comment by ryzokuken** [comment]
@jywarren not even close to 4

https://github.com/publiclab/plots2/blob/d5da035c98f64d5ad0f51ea463e6a27c802dda67/Gemfile#L32

**Comment by bellvat** [comment]
@ryzokuken Sure. 
Just to confirm, I would add `:branch => 'master'` to the end of the ruby-prof line in the Gemfile, then run `bundle update`, correct?

**Comment by bellvat** [comment]
@jywarren and @ryzokuken 
I am going to try to install an earlier version of ruby-prof to see if it will help. If not, we can revert to an earlier commit.

**Comment by ryzokuken** [comment]
No, I don't think that'll help. Which ruby version are you using BTW?

**Comment by bellvat** [comment]
@ryzokuken  
Got it. Will revert back. Ruby version 2.3.4.

**Comment by ryzokuken** [comment]
Yeah, it might help. If it doesn't also try mentioning the `master` branch in your Gemfile and run `bundle update` everytime you make a change in the Gemfile.

**Comment by jywarren** [comment]
For uglifier, `>=` is anything equal or higher, while `~>` is anything within a `0.0.x` bugfix/patch level change. So if you don't want it to go up to 4, we'd want to change it to `~>3.0.0` (i'm not sure if there's a way to specify a `0.x` level change but not a major version # change)

Do you think `uglifier` is the problem? If so, shall we limit it to v3? I still see the error:

`/usr/local/bundle/bundler/gems/ruby-prof-d6867cf1133b/lib/ruby-prof/printers/call_tree_printer.rb:46:in `print': wrong number of arguments (given 2, expected 0..1) (ArgumentError)`

And we seem to be on `ruby-prof` version `0.16.2`

**Comment by jywarren** [comment]
I'm sorry this one has turned out to be so much trouble, thanks so much for your persistence! 

I think there's a typo in this last test, let me see.

**Comment by jywarren** [comment]
It's a weird error;

`ERROR: Service 'web' failed to build: failed to process "\"This": unexpected end of statement while looking for matching double-quote`

I restarted the test to see if it was a fluke. Fingers X!

**Comment by ryzokuken** [comment]
@jywarren Did we specify the master branch yet? `v0.16.2` doesn't work for us.

**Comment by jywarren** [comment]
I think so - i see branch => master

sorry, what version should we specify?

**Comment by bellvat** [comment]
Hi @jywarren  and @ryzokuken,

I made the following changes in the latest commit:
1. I reverted the branch, so currently, versions for existing gems match masters.
2. I am using Ruby version 2.3.4
3. I found that when running `rake test:all` on the Master branch, I see that 577 tests pass, but 7/7 solr tests are erroneous- please see the following: 
screen shot 2017-12-12 at 9 20 12 am
screen shot 2017-12-12 at 9 20 31 am
screen shot 2017-12-12 at 9 20 47 am
screen shot 2017-12-12 at 9 21 16 am

4. When I ran `rake test:all` with ruby-prof version 0.16.2 on the Performance-test branch, I got the following message: 
screen shot 2017-12-12 at 9 48 07 am

5. When I switched to ruby-prof version 0.15.9 on the Performance-test branch, `rake test:benchmark` passes, as follows:
screen shot 2017-12-12 at 9 13 13 am

6. Then when I run `rake test:all` on the Performance-test branch with ruby-prof(0.15.9), I see that 577 tests pass, and get the same 7/7 solr test errors like in the Master branch. Please see below:
screen shot 2017-12-12 at 9 16 14 am
screen shot 2017-12-12 at 9 07 57 am
screen shot 2017-12-12 at 9 08 22 am
screen shot 2017-12-12 at 9 09 05 am

It might be something about my environment that is causing this?
Would it be worthwhile for one of you to try adding the ruby-prof gem and uncommenting the browser_test.rb on your local to give it a try?

**Comment by jywarren** [comment]
Hmm, I wonder if the new error `: Service 'web' failed to build: failed to process "\"This": unexpected end of statement while looking for matching double-quote` is now blocking, and the original version bug is actually resolved now. I'm wondering where we could possibly have missed a double quote...

**Comment by jywarren** [comment]
Literally the only double quote in your pull requests is `test "browsers" do` -- it sounds wild, but maybe changing that to single quotes could have some effect?

**Comment by jywarren** [comment]
OK I think @ryzokuken solved the unrelated issue -- and i'm rerunning the tests now to see if they pass. I think they should now run on top of the change in #1881 -- thanks everyone for your help and patience!

**Comment by jywarren** [comment]
Hmm, maybe we have to push some new commit here. I still see the old Dockerfile:

https://github.com/bellvat/plots2/blob/c6b28d54097ca6dc6d3aae88af27fc2a99b903fa/Dockerfile#L7

Let me try pushing something insignificant.

**Comment by jywarren** [comment]
I went ahead and rebased all these changes on top of the latest, just to be sure.

**Comment by ryzokuken** [comment]
@jywarren It works! All we now need to do is find a way to make the reporter parse this test. I'm taking a look at https://github.com/ruby-prof/ruby-prof#printers

**Comment by ryzokuken** [comment]
@jywarren Can we just exclude it from report generation? AFAIK, Rails Prof can generate it's own reports in multiple formats.

**Comment by jywarren** [comment]
For now, sure. I think it'd be great to have Dangerbot report the effects on performance of different PRs -- but we can make a follow-up issue for that!

**Comment by jywarren** [comment]
How might it be disabled?

**Comment by jywarren** [comment]
Hi, @ryzokuken - did you have an idea of how to disable report generation
for performance tests? Your help is MUCH appreciated, thank you!

**Comment by jywarren** [comment]
Happy new year, all! @ryzokuken did you have an idea of how we could disable report parsing, and just output the rails-perf native report? The error is currently when junit tries to parse the report here:

`/usr/local/bundle/gems/minitest-reporters-1.1.19/lib/minitest/reporters/junit_reporter.rb:114:in `+': nil can't be coerced into Fixnum (TypeError)`

Or we could try running this locally and see what format the report is in, and see why there's a `nil` in there...

**Comment by jywarren** [comment]
Interestingly it's failing on the "time" line in this function of the `junit_reporter.rb` file:

```ruby

      def analyze_suite(tests)
        result = Hash.new(0)
        tests.each do |test|
          result[:"#{result(test)}_count"] += 1
          result[:assertion_count] += test.assertions
          result[:test_count] += 1
          result[:time] += test.time
        end
        result
      end
```

**Comment by jywarren** [comment]
Well, this is good!  I get the same error when running this locally.

**Comment by jywarren** [comment]
Some clues in the `ruby-prof` docs say we need to add a Rack adapter to `config.ru` --

```
if Rails.env.profile?
  use Rack::RubyProf, :path => '/temp/profile'
end
```

ugggg but it'd be `Rails.env.test?`

**Comment by jywarren** [comment]
Ooh, a lot of clues here: http://ccaloha.cc/blog/2015/07/14/howto-performance-test-in-rails-4/

and here: http://railscasts.com/episodes/411-performance-testing

**Comment by jywarren** [comment]
I'm a bit stuck, to be honest. Performance testing docs in Rails 3.2: http://guides.rubyonrails.org/v3.2/testing.html#rake-tasks-for-running-your-tests are missing from Rails 4.1: http://guides.rubyonrails.org/v4.1/testing.html

I guess we need to get junit to ignore rails-perftest. I'm going to use some of the options that were commented out in the test...

**Comment by jywarren** [comment]
No luck! The error is... odd. Especially since I didn't get it when we first started! I think it may have to do with us now being on Rails 4.1, but not sure.

**Comment by jywarren** [comment]
Trying to echo out the reports with `cat test/reports/*` -- see if we can see inside a bit!

**Comment by jywarren** [comment]
That was instructive -- i believe it's not generating a report at all! @ryzokuken what do you think?

**Comment by ryzokuken** [comment]
@jywarren let me check.

**Comment by jywarren** [comment]
@publiclab/reviewers -- this is a real tough one. If anyone can give it a try to see if we can move this forward, much appreciated! I'm totally stuck!

**Comment by ryzokuken** [comment]
@jywarren me too, tried a couple of times.

**Comment by jywarren** [comment]
OK, i'm going to try this again! Fingers X -- pushing a newly rebased version up!

**Comment by jywarren** [comment]
OK, a clarification -- this seems to be happening when /generating/ the junit reports, but before reading and parsing them in the Dangerfile for report-back in the PR.

**Comment by jywarren** [comment]
I disabled the offending line of minitest-reporters -- 113 -- and got it to run. THe result:

```
(xenial)warren@localhost:~/sites/plots2$ less test/reports/
BrowsingTest#test_browsers_memory.csv          BrowsingTest#test_browsers_wall_time.csv       TEST-BrowsingTest.xml
BrowsingTest#test_browsers_memory_flat.txt     BrowsingTest#test_browsers_wall_time_flat.txt  
(xenial)warren@localhost:~/sites/plots2$ less test/reports/TEST-BrowsingTest.xml 
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="BrowsingTest" skipped="0" failures="0" errors="0" tests="1" assertions="0" time="0">
  <testcase name="test_browsers" classname="BrowsingTest" assertions="0" time="">
  </testcase>
</testsuite>
```

**Comment by jywarren** [comment]
OK, I think we're going to have to give up on this for now, unless the folks over at https://github.com/kern/minitest-reporters/issues/251 are able to help us out at all; 

For anyone trying to solve this, the tests are running, but the XML JUnit reports produced by `minitest-reporters` are failing to get generated due to line 114 of this file:

https://github.com/kern/minitest-reporters/blob/9d3f5777c1a45e576a0c6844fa1f612831768595/lib/minitest/reporters/junit_reporter.rb#L114

@bellvat I'm so sorry we weren't able to make it on this! I hope it wasn't a terrible experience for your first contribution and I'd love to help you choose another over at https://publiclab.github.io/community-toolbox/#r=all if you're still interested. Many thanks for working with us on this one that turned out to be unexpectedly awful!

**Comment by bellvat** [comment]
@jywarren No problem! I wish I could have helped out more with the reporting issues. Thanks for really trying to work this out!

**Comment by ryzokuken** [comment]
@jywarren keep this open, I'll keep trying to fix this somehow as my understanding of our test reporting improves.

**Comment by jywarren** [comment]
Thanks, all!

**Comment by jywarren** [comment]
Whoa, they just adopted our solution in `minitest-reporters` -- look! https://github.com/kern/minitest-reporters/issues/251

`minitest-reporters v1.2.0.beta2` might now work for us!

**Comment by jywarren** [comment]
I just tried updating minitest-reporters to `v1.2.0.beta2` to see if that passes.

**Comment by jywarren** [comment]
Ooh! A new error!

`/usr/local/bundle/gems/minitest-reporters-1.2.0.beta2/lib/minitest/reporters/junit_reporter.rb:56:in `method': undefined method `test_newcomer_should_get_legacy_form' for class `Minitest::Result' (NameError)`

**Comment by jywarren** [comment]
That's odd. It's for this test, of all things: 

https://github.com/publiclab/plots2/blob/5ffd7c6b3f9d2a3a53946a3433970ce1590ceb15/test/functional/editor_controller_test.rb#L71

I can't see why that test buried in the middle of our tests should have anything to do with performance tests. I'll try running this again...?

**Comment by jywarren** [comment]
Unfortunately, [we've learned](https://github.com/publiclab/plots2/pull/2726#issuecomment-394833053) via @siaw23 that Rails performance testing was mostly for Rails 3 and has been supported less and less. I think he's right that we should close this and pursue something with #2660 -- Skylake -  Thanks @siaw23 !!!

And thanks @bellvat for your help with this and I'm sorry we never got this attempt running, but it looks like it was none of our faults!

**Pull request closed without merging**